
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Kernels &#8212; NDD II - Investigating Kernels</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="shortcut icon" href="_static/nd_logo_small.png"/>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Matched Networks" href="mice_discriminability.html" />
    <link rel="prev" title="Datasets" href="data_description.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/nd_logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">NDD II - Investigating Kernels</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Abstract
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="data_description.html">
   Datasets
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Kernels
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Discriminability
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="mice_discriminability.html">
   Matched Networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mami_discriminability.html">
   Unmatched Networks
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Clustering
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="mice_clustering.html">
   Matched Networks
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mami_clustering.html">
   Unmatched Networks
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Conclusion
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="conclusion.html">
   Conclusion
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Appendix
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="laplacian.html">
   Working on Laplacian Kernel
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="mami_replicatefigs.html">
   Replicating Figures from Paper
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/kernels.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/neurodata/ndd-dlee"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/neurodata/ndd-dlee/issues/new?title=Issue%20on%20page%20%2Fkernels.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/neurodata/ndd-dlee/main?urlpath=tree/docs/kernels.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Kernels
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#for-matched-networks">
   For Matched Networks
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#global-properties">
     Global Properties
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#density">
       Density
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#average-edge-weight">
       Average Edge Weight
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#average-of-the-adjacency-matrix">
       Average of the Adjacency Matrix
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#node-wise-properties">
     Node-wise Properties
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#node-degree">
       Node Degree
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#node-strength">
       Node Strength
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#edge-wise-properties">
     Edge-wise Properties
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#edge-weight">
       Edge Weight
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#for-unmatched-networks">
   For Unmatched Networks
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#established-kernels">
   Established Kernels
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#laplacian-spectral-distance">
     Laplacian Spectral Distance
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#for-matched-networks-omnibus-embedding">
     For Matched Networks - Omnibus Embedding
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#for-unmatched-networks-latent-distribution-test-statistic">
     For Unmatched Networks - Latent Distribution Test Statistic
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   References
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Kernels</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Kernels
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#for-matched-networks">
   For Matched Networks
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#global-properties">
     Global Properties
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#density">
       Density
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#average-edge-weight">
       Average Edge Weight
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#average-of-the-adjacency-matrix">
       Average of the Adjacency Matrix
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#node-wise-properties">
     Node-wise Properties
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#node-degree">
       Node Degree
      </a>
     </li>
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#node-strength">
       Node Strength
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#edge-wise-properties">
     Edge-wise Properties
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#edge-weight">
       Edge Weight
      </a>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#for-unmatched-networks">
   For Unmatched Networks
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#established-kernels">
   Established Kernels
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#laplacian-spectral-distance">
     Laplacian Spectral Distance
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#for-matched-networks-omnibus-embedding">
     For Matched Networks - Omnibus Embedding
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#for-unmatched-networks-latent-distribution-test-statistic">
     For Unmatched Networks - Latent Distribution Test Statistic
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   References
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="kernels">
<h1>Kernels<a class="headerlink" href="#kernels" title="Permalink to this headline">¶</a></h1>
<p>We build our simple kernels using global, node-wise, and edge-wise properties of the network. Both datasets have the shape <span class="math notranslate nohighlight">\((N, n, n)\)</span> where <span class="math notranslate nohighlight">\(N\)</span> is the number of networks in the dataset and <span class="math notranslate nohighlight">\(n\)</span> is the number of nodes in each network. Each graph is represented by an <span class="math notranslate nohighlight">\((n\)</span> x <span class="math notranslate nohighlight">\(n)\)</span> adjacency matrix, where the existence of an edge between node <span class="math notranslate nohighlight">\(i\)</span> and node <span class="math notranslate nohighlight">\(j\)</span> is denoted by a nonzero value at the <span class="math notranslate nohighlight">\(i,j\)</span> th element.</p>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="for-matched-networks">
<h1>For Matched Networks<a class="headerlink" href="#for-matched-networks" title="Permalink to this headline">¶</a></h1>
<p>Attached is the code for calculating the dissimilarity matrix using different kernels for matched networks.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># HIDE CODE</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="k">def</span> <span class="nf">calculate_dissim</span><span class="p">(</span><span class="n">graphs</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;density&quot;</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Calculate the dissimilarity matrix using the input kernel. &quot;&quot;&quot;</span>
    <span class="n">glob</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">node</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="n">edge</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;density&quot;</span><span class="p">:</span>
        <span class="n">glob</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">num_nodes</span> <span class="o">=</span> <span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
        <span class="n">num_edges_possible</span> <span class="o">=</span> <span class="n">num_nodes</span> <span class="o">**</span> <span class="mi">2</span> <span class="o">-</span> <span class="n">num_nodes</span>

        <span class="n">metric</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
            <span class="n">num_edges</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
            <span class="n">metric</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">num_edges</span> <span class="o">/</span> <span class="n">num_edges_possible</span>
    
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;avgedgeweight&quot;</span><span class="p">:</span>
        <span class="n">glob</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">metric</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
            <span class="n">num_edges</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
            <span class="n">metric</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span> <span class="o">/</span> <span class="n">num_edges</span>
    
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;avgadjmatrix&quot;</span><span class="p">:</span>
        <span class="n">glob</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">metric</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
            <span class="n">metric</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">average</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>   

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;degree&quot;</span><span class="p">:</span>
        <span class="n">node</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">metric</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graph</span><span class="p">):</span>
                <span class="n">metric</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;strength&quot;</span><span class="p">:</span>
        <span class="n">node</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">metric</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graph</span><span class="p">):</span>
                <span class="n">metric</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;edgeweight&quot;</span><span class="p">:</span>
        <span class="n">edge</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="n">metric</span> <span class="o">=</span> <span class="n">graphs</span>
    
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Not a valid kernel name.&quot;</span><span class="p">)</span>
    
    <span class="n">dissim_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">)))</span>

    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">metric1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">metric</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">metric2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">metric</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">glob</span> <span class="ow">and</span> <span class="n">norm</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">metric1</span><span class="o">-</span> <span class="n">metric2</span><span class="p">)</span>
            <span class="k">elif</span> <span class="p">(</span><span class="n">node</span> <span class="ow">or</span> <span class="n">edge</span><span class="p">)</span> <span class="ow">and</span> <span class="n">norm</span> <span class="o">==</span> <span class="s2">&quot;l1&quot;</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">metric1</span> <span class="o">-</span> <span class="n">metric2</span><span class="p">,</span> <span class="nb">ord</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">elif</span> <span class="p">(</span><span class="n">node</span> <span class="ow">or</span> <span class="n">edge</span><span class="p">)</span> <span class="ow">and</span> <span class="n">norm</span> <span class="o">==</span> <span class="s2">&quot;l2&quot;</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">metric1</span> <span class="o">-</span> <span class="n">metric2</span><span class="p">,</span> <span class="nb">ord</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;L1, L2 norms only apply to node or edge-wise kernels.&quot;</span><span class="p">)</span>
            
            <span class="n">dissim_matrix</span><span class="p">[</span><span class="n">j</span><span class="p">,</span> <span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">diff</span>

    <span class="k">if</span> <span class="n">normalize</span><span class="p">:</span>
        <span class="n">dissim_matrix</span> <span class="o">=</span> <span class="n">dissim_matrix</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">dissim_matrix</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">dissim_matrix</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="global-properties">
<h2>Global Properties<a class="headerlink" href="#global-properties" title="Permalink to this headline">¶</a></h2>
<p>For each of these kernels, each element of the <span class="math notranslate nohighlight">\((N\)</span> x <span class="math notranslate nohighlight">\(N)\)</span> dissimilarity matrix is the absolute difference between the metrics, where <span class="math notranslate nohighlight">\(N\)</span> is the number of networks in the dataset.</p>
<div class="section" id="density">
<h3>Density<a class="headerlink" href="#density" title="Permalink to this headline">¶</a></h3>
<p>Graph density refers to the ratio between the number of edges present in the graph and the number of possible edges and is calculated by</p>
<div class="math notranslate nohighlight">
\[    \text{Density} = \frac{\text{Number of Edges Present}}{n(n-1)}\]</div>
<p>where <span class="math notranslate nohighlight">\(n\)</span> is the number of nodes in the graph.</p>
</div>
<div class="section" id="average-edge-weight">
<h3>Average Edge Weight<a class="headerlink" href="#average-edge-weight" title="Permalink to this headline">¶</a></h3>
<p>To calculate average edge weight, we sum all the nonzero values in the adjacency matrix and divide by the number of edges present.</p>
<div class="math notranslate nohighlight">
\[    \text{Average Edge Weight} = \frac{\text{Sum of Edge Weights}}{\text{Number of Edges Present}}\]</div>
</div>
<div class="section" id="average-of-the-adjacency-matrix">
<h3>Average of the Adjacency Matrix<a class="headerlink" href="#average-of-the-adjacency-matrix" title="Permalink to this headline">¶</a></h3>
<p>To calculate the average of the adjacency matrix, we sum all the nonzero values in the adjacency matrix and divide by the total number of possible edges. Note that if the graph is unweighted, the average of the adjacency matrix would be the same as graph density.</p>
<div class="math notranslate nohighlight">
\[    \text{Average of Adj. Matrix} = \frac{\text{Sum of Edge Weights}}{n(n-1)}\]</div>
<p>where <span class="math notranslate nohighlight">\(n\)</span> is the number of nodes in the graph.</p>
</div>
</div>
<div class="section" id="node-wise-properties">
<h2>Node-wise Properties<a class="headerlink" href="#node-wise-properties" title="Permalink to this headline">¶</a></h2>
<p>For each of these kernels, two <span class="math notranslate nohighlight">\((N\)</span> x <span class="math notranslate nohighlight">\(N)\)</span> dissimilarity matrices are constructed. Each element of the dissimilarity matrix is the <span class="math notranslate nohighlight">\(L^1\)</span>-norm and <span class="math notranslate nohighlight">\(L^2\)</span>-norm of the difference vector between the metrics.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <span class="math notranslate nohighlight">\(L^1\)</span>-norm is the sum of the magnitude of the vector. Formally,</p>
<div class="math notranslate nohighlight">
\[    |\boldsymbol{x}|_1 = \sum_{i=1}^n |x_i|\]</div>
<p>where <span class="math notranslate nohighlight">\(x_i\)</span> for <span class="math notranslate nohighlight">\(i = 1, 2, ..., n\)</span> are the elements of the vector <span class="math notranslate nohighlight">\(\boldsymbol{x}\)</span> <span id="id1">[<a class="reference internal" href="#id10" title="Eric W. Weisstein. L^1-norm. Accessed May 16, 2022 [Online]. URL: https://mathworld.wolfram.com/L1-Norm.html.">Weisstein, n.d.</a>]</span>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <span class="math notranslate nohighlight">\(L^2\)</span>-norm is the Euclidean distance from the origin to the vector. Formally,</p>
<div class="math notranslate nohighlight">
\[    |\boldsymbol{x}|_2 = \sqrt{\sum_{i=1}^n |x_i|^2}\]</div>
<p>where <span class="math notranslate nohighlight">\(x_i\)</span> for <span class="math notranslate nohighlight">\(i = 1, 2, ..., n\)</span> are the elements of the vector <span class="math notranslate nohighlight">\(\boldsymbol{x}\)</span> <span id="id2">[<a class="reference internal" href="#id11" title="Eric W. Weisstein. L^2-norm. Accessed May 16, 2022 [Online]. URL: https://mathworld.wolfram.com/L2-Norm.html.">Weisstein, n.d.</a>]</span>.</p>
</div>
<div class="section" id="node-degree">
<h3>Node Degree<a class="headerlink" href="#node-degree" title="Permalink to this headline">¶</a></h3>
<p>The degree of a node is the number of edges a node has. In an <span class="math notranslate nohighlight">\((n\)</span> x <span class="math notranslate nohighlight">\(n)\)</span> adjacency matrix <span class="math notranslate nohighlight">\(A\)</span>, the degree of node <span class="math notranslate nohighlight">\(i\)</span> is calculated by <span class="math notranslate nohighlight">\(\sum_{j=1}^n a_{j}\)</span> where <span class="math notranslate nohighlight">\(a_{j}=1\)</span> if <span class="math notranslate nohighlight">\(A_{ij} \neq 0\)</span> and <span class="math notranslate nohighlight">\(a_{j}=0\)</span> if <span class="math notranslate nohighlight">\(A_{ij} = 0\)</span>. Thus for each graph, we have an <span class="math notranslate nohighlight">\((i\)</span> x <span class="math notranslate nohighlight">\(1)\)</span> vector where each element is the degree of node <span class="math notranslate nohighlight">\(i\)</span>, and the dissimilarity matrix is constructed based on these vectors.</p>
</div>
<div class="section" id="node-strength">
<h3>Node Strength<a class="headerlink" href="#node-strength" title="Permalink to this headline">¶</a></h3>
<p>The strength of a node is the sum of the weights of the edges a node has. In an <span class="math notranslate nohighlight">\((n\)</span> x <span class="math notranslate nohighlight">\(n)\)</span> adjacency matrix <span class="math notranslate nohighlight">\(A\)</span>, the degree of node <span class="math notranslate nohighlight">\(i\)</span> is calculated by <span class="math notranslate nohighlight">\(\sum_{j=1}^n A_{ij}\)</span>. Thus for each graph, we have an <span class="math notranslate nohighlight">\((i\)</span> x <span class="math notranslate nohighlight">\(1)\)</span> vector where each element is the strength of node <span class="math notranslate nohighlight">\(i\)</span>, and the dissimilarity matrix is constructed based on these vectors.</p>
</div>
</div>
<div class="section" id="edge-wise-properties">
<h2>Edge-wise Properties<a class="headerlink" href="#edge-wise-properties" title="Permalink to this headline">¶</a></h2>
<p>Similar to the kernels above, two <span class="math notranslate nohighlight">\((N\)</span> x <span class="math notranslate nohighlight">\(N)\)</span> dissimilarity matrices are constructed based on the <span class="math notranslate nohighlight">\(L^1\)</span>-matrix norm and <span class="math notranslate nohighlight">\(L^2\)</span>-matrix norm of the difference matrix between the metrics.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <span class="math notranslate nohighlight">\(L^1\)</span>-matrix norm is also called the maximum absolute column sum norm and is calculated by</p>
<div class="math notranslate nohighlight">
\[    ||\boldsymbol{A}_1|| = \text{max}_j \sum_{i=1}^n |A_ij|\]</div>
<p><span id="id3">[<a class="reference internal" href="#id12" title="Eric W. Weisstein. Matrix norm. Accessed May 16, 2022 [Online]. URL: https://mathworld.wolfram.com/MatrixNorm.html.">Weisstein, n.d.</a>]</span>.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <span class="math notranslate nohighlight">\(L^2\)</span>-matrix norm is also called the spectral norm and is calculated by</p>
<div class="math notranslate nohighlight">
\[    ||\boldsymbol{A}_2|| = (\text{max eigenvalue of} A^H A)^{1/2}\]</div>
<p>where <span class="math notranslate nohighlight">\(A^H\)</span> is the conjugate transpose matrix. In our case, matrix <span class="math notranslate nohighlight">\(A\)</span> is an adjacency matrix and thus always real, so <span class="math notranslate nohighlight">\(A^H = A^T\)</span> <span id="id4">[<a class="reference internal" href="#id12" title="Eric W. Weisstein. Matrix norm. Accessed May 16, 2022 [Online]. URL: https://mathworld.wolfram.com/MatrixNorm.html.">Weisstein, n.d.</a>]</span>.</p>
</div>
<div class="section" id="edge-weight">
<h3>Edge Weight<a class="headerlink" href="#edge-weight" title="Permalink to this headline">¶</a></h3>
<p>For each pair of <span class="math notranslate nohighlight">\((n\)</span> x <span class="math notranslate nohighlight">\(n)\)</span> graphs, the difference in edge weights is calculated by the difference in their adjacency matrices, yielding an <span class="math notranslate nohighlight">\((n\)</span> x <span class="math notranslate nohighlight">\(n\)</span>) difference matrix. Then the dissimilarity matrix is constructed based on the <span class="math notranslate nohighlight">\(L^1\)</span> and <span class="math notranslate nohighlight">\(L^2\)</span>-matrix norm of these matrices.</p>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="for-unmatched-networks">
<h1>For Unmatched Networks<a class="headerlink" href="#for-unmatched-networks" title="Permalink to this headline">¶</a></h1>
<p>The kernels we use above assume the nodes are matched. However, when we have no known matching of the nodes, we cannot use kernels based on the node-wise or edge-wise properties of the network because we cannot calculate a difference vector or matrix directly and calculate its <span class="math notranslate nohighlight">\(L^1\)</span> and <span class="math notranslate nohighlight">\(L^2\)</span> norm. Thus, we flatten the difference vector or matrix for those kernels and find the <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ks_2samp.html">two-sample Kolmogorov-Smirnov test statistic</a> for the two “distributions”. The metrics remain the same. The modified code is attached below.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># HIDE CODE</span>
<span class="kn">from</span> <span class="nn">scipy.stats</span> <span class="kn">import</span> <span class="n">ks_2samp</span>

<span class="k">def</span> <span class="nf">calculate_dissim_unmatched</span><span class="p">(</span><span class="n">graphs</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;degree&quot;</span><span class="p">,</span> <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot; Calculate the dissimilarity matrix using the input kernel. &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;degree&quot;</span><span class="p">:</span>
        <span class="n">metric</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graph</span><span class="p">):</span>
                <span class="n">metric</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>
    
    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;strength&quot;</span><span class="p">:</span>
        <span class="n">metric</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">graphs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">node</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graph</span><span class="p">):</span>
                <span class="n">metric</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">node</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">method</span> <span class="o">==</span> <span class="s2">&quot;edgeweight&quot;</span><span class="p">:</span>
        <span class="n">metric</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">graph</span> <span class="ow">in</span> <span class="n">graphs</span><span class="p">:</span>
            <span class="n">metric</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ravel</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">graph</span><span class="p">)))</span>
            
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Not a valid kernel name.&quot;</span><span class="p">)</span>
    
    <span class="n">dissim_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">)))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">metric1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">metric</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">metric2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">metric</span><span class="p">):</span>
            <span class="n">diff</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">ks_2samp</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">metric1</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">metric2</span><span class="p">),</span> <span class="n">mode</span><span class="o">=</span><span class="s1">&#39;asymp&#39;</span><span class="p">)</span>
            <span class="n">dissim_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">diff</span>
            
    <span class="k">if</span> <span class="n">normalize</span><span class="p">:</span>
        <span class="n">dissim_matrix</span> <span class="o">=</span> <span class="n">dissim_matrix</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">dissim_matrix</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">dissim_matrix</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="established-kernels">
<h1>Established Kernels<a class="headerlink" href="#established-kernels" title="Permalink to this headline">¶</a></h1>
<p>The following kernels use existing algorithms and/or equations that are much more complex and computationally expensive than those described above. We use them as a point of comparison for the performance for our “simple” kernels.</p>
<div class="section" id="laplacian-spectral-distance">
<h2>Laplacian Spectral Distance<a class="headerlink" href="#laplacian-spectral-distance" title="Permalink to this headline">¶</a></h2>
<p>We compute the <a class="reference external" href="https://microsoft.github.io/graspologic/latest/reference/reference/utils.html#graspologic.utils.to_laplacian">I-DAD Laplacian</a> of each graph, which is calculated by</p>
<div class="math notranslate nohighlight">
\[L = I - D_i A D_i\]</div>
<p>where <span class="math notranslate nohighlight">\(D\)</span> is the diagonal matrix of degrees of node <span class="math notranslate nohighlight">\(i\)</span> raised to the -1/2 power, <span class="math notranslate nohighlight">\(I\)</span> is the identity matrix, and <span class="math notranslate nohighlight">\(A\)</span> is the adjacency matrix.</p>
<p>Graphs can often be characterized by its Laplacian eigenspectrum, which makes this a useful metric to consider when assessing similarity between graphs <span id="id5">[<a class="reference internal" href="#id14" title="K. Ch. Das. The laplacian spectrum of a graph. Computers &amp; Mathematics with Applications, 48(5-6):715-724, 09 2004. doi:10.1016/j.camwa.2004.05.005.">Das, 2004</a>]</span>. The I-DAD Laplacian is particularly useful because all its eigenvalues are real, and for undirected graphs, all eigenvalues are guaranteed to be in the range [0, 2] <span id="id6">[<a class="reference internal" href="mami_discriminability.html#id3" title="Laura E Suárez, Yossi Yovel, Martijin P Van den Heuvel, Olaf Sporns, Yaniv Assaf, Guillaume Lajoie, and Bratislav Misic. A connectomics-based taxonomy of mammals. bioRxiv, 03 2022. doi:10.1101/2022.03.11.483995.">Suárez <em>et al.</em>, 2022</a>]</span>.</p>
<p>We implement the Laplacian spectral kernel by applying pass-to-ranks transformations on the networks, sorting the eigenvalues of the graph Laplacian without any smoothing, then calculating the <span class="math notranslate nohighlight">\(L^2\)</span>-norm of the difference vector to construct the dissimilarity matrix.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Using the code attached below, there are several ways we can derive and use the Laplacian eigenspectrum. We can transform the graphs through <a class="reference external" href="https://microsoft.github.io/graspologic/latest/reference/reference/utils.html#graspologic.utils.pass_to_ranks">pass-to-ranks</a> or binarization before finding its graph Laplacian, and we can simply sort its eigenvalues or smooth them using a Gaussian kernel to construct the eigenspectrum. We also have multiple ways to quantify dissimilarity, such as calculating the <a class="reference external" href="https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.cosine.html">cosine difference</a> between two eigenspectra, or by finding the <span class="math notranslate nohighlight">\(L^1\)</span>-norm, or <span class="math notranslate nohighlight">\(L^2\)</span>-norm of the difference vector between two eigenspectra. The implementation we state above is based on which yields the highest discriminiability value on the matched networks. More details can be found in the <span class="xref std std-doc">Laplacian</span> section in the Appendix.</p>
</div>
<p>We use the Laplacian spectral kernel for both matched and unmatched networks.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># HIDE CODE</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">graspologic.utils</span> <span class="kn">import</span> <span class="n">pass_to_ranks</span><span class="p">,</span> <span class="n">to_laplacian</span>
<span class="kn">from</span> <span class="nn">sklearn.neighbors</span> <span class="kn">import</span> <span class="n">KernelDensity</span>
<span class="kn">from</span> <span class="nn">scipy.spatial.distance</span> <span class="kn">import</span> <span class="n">cosine</span>

<span class="k">def</span> <span class="nf">laplacian_dissim</span><span class="p">(</span><span class="n">graphs</span><span class="p">,</span> <span class="n">transform</span><span class="p">:</span> <span class="nb">str</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">metric</span><span class="p">:</span> <span class="nb">str</span><span class="o">=</span><span class="s1">&#39;l2&#39;</span><span class="p">,</span> <span class="n">smooth_eigvals</span><span class="p">:</span> <span class="nb">bool</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> \
    <span class="n">normalize</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">transform</span> <span class="o">==</span> <span class="s1">&#39;pass-to-ranks&#39;</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
            <span class="n">graph</span> <span class="o">=</span> <span class="n">pass_to_ranks</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
            <span class="n">graphs</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">graph</span>
    <span class="k">elif</span> <span class="n">transform</span> <span class="o">==</span> <span class="s1">&#39;binarize&#39;</span><span class="p">:</span>
        <span class="n">graphs</span><span class="p">[</span><span class="n">graphs</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">elif</span> <span class="n">transform</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">graphs</span> <span class="o">=</span> <span class="n">graphs</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Supported transformations are &quot;pass-to-ranks&quot; (simple-nonzero), &quot;binarize&quot;, or None.&#39;</span><span class="p">)</span>
    
    <span class="n">eigs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
        <span class="c1"># calculate laplacian</span>
        <span class="n">lap</span> <span class="o">=</span> <span class="n">to_laplacian</span><span class="p">(</span><span class="n">graph</span><span class="p">,</span> <span class="s1">&#39;I-DAD&#39;</span><span class="p">)</span>

        <span class="c1"># find and sort eigenvalues</span>
        <span class="n">w</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">eigvals</span><span class="p">(</span><span class="n">lap</span><span class="p">)</span>
        <span class="n">w</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">w</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">smooth_eigvals</span><span class="p">:</span>
            <span class="n">kde</span> <span class="o">=</span> <span class="n">KernelDensity</span><span class="p">(</span><span class="n">kernel</span><span class="o">=</span><span class="s1">&#39;gaussian&#39;</span><span class="p">,</span> <span class="n">bandwidth</span><span class="o">=</span><span class="mf">0.015</span><span class="p">)</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">w</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
            <span class="n">xs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2000</span><span class="p">)</span>
            <span class="n">xs</span> <span class="o">=</span> <span class="n">xs</span><span class="p">[:,</span> <span class="n">np</span><span class="o">.</span><span class="n">newaxis</span><span class="p">]</span>
            <span class="n">log_dens</span> <span class="o">=</span> <span class="n">kde</span><span class="o">.</span><span class="n">score_samples</span><span class="p">(</span><span class="n">xs</span><span class="p">)</span>
            <span class="n">eigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">exp</span><span class="p">(</span><span class="n">log_dens</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">eigs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">w</span><span class="p">)</span>

    <span class="n">dissim_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">)))</span>
    <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">eig1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">eigs</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">eig2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">eigs</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;cosine&#39;</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">cosine</span><span class="p">(</span><span class="n">eig1</span><span class="p">,</span> <span class="n">eig2</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;l1&#39;</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">eig1</span> <span class="o">-</span> <span class="n">eig2</span><span class="p">,</span> <span class="nb">ord</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="k">elif</span> <span class="n">metric</span> <span class="o">==</span> <span class="s1">&#39;l2&#39;</span><span class="p">:</span>
                <span class="n">diff</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">eig1</span> <span class="o">-</span> <span class="n">eig2</span><span class="p">,</span> <span class="nb">ord</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
            <span class="n">dissim_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">diff</span>

    <span class="k">if</span> <span class="n">normalize</span><span class="p">:</span>
        <span class="n">dissim_matrix</span> <span class="o">=</span> <span class="n">dissim_matrix</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">dissim_matrix</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">dissim_matrix</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="for-matched-networks-omnibus-embedding">
<h2>For Matched Networks - Omnibus Embedding<a class="headerlink" href="#for-matched-networks-omnibus-embedding" title="Permalink to this headline">¶</a></h2>
<p>We generate <a class="reference external" href="https://microsoft.github.io/graspologic/latest/reference/reference/embed.html#graspologic.embed.OmnibusEmbed">omnibus embeddings</a> of the matched networks, then calculate the Frobenius norm of the difference between two embeddings to construct the dissimilarity matrix.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The Frobenius norm of a matrix <span class="math notranslate nohighlight">\(A\)</span> is given by</p>
<div class="math notranslate nohighlight">
\[||A||_F = (\sum_{i,j} |A_{ij}|^2)^{1/2}\]</div>
<p><span id="id7">[<a class="reference internal" href="#id13" title="Gene H. Golub and Charles F. Van Loan. Matrix Computations. The Johns Hopkins University Press, Baltimore, MD, 1985.">Golub and Van Loan, 1985</a>]</span>.</p>
</div>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># HIDE CODE</span>
<span class="kn">from</span> <span class="nn">graspologic.embed</span> <span class="kn">import</span> <span class="n">OmnibusEmbed</span>

<span class="c1"># embed using Omni</span>
<span class="n">embedder</span> <span class="o">=</span> <span class="n">OmnibusEmbed</span><span class="p">(</span><span class="n">n_elbows</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<span class="n">omni_embedding</span> <span class="o">=</span> <span class="n">embedder</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">graphs</span><span class="p">)</span>

<span class="n">omni_matrix</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">graphs</span><span class="p">)))</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">embedding1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">omni_embedding</span><span class="p">):</span>
    <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">embedding2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">omni_embedding</span><span class="p">):</span>
        <span class="n">dist</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linalg</span><span class="o">.</span><span class="n">norm</span><span class="p">(</span><span class="n">embedding1</span> <span class="o">-</span> <span class="n">embedding2</span><span class="p">,</span> <span class="nb">ord</span><span class="o">=</span><span class="s2">&quot;fro&quot;</span><span class="p">)</span>
        <span class="n">omni_matrix</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist</span>

<span class="c1"># scale values from 0 to 1</span>
<span class="n">scaled_omni_dissim</span> <span class="o">=</span> <span class="n">omni_matrix</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">omni_matrix</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="for-unmatched-networks-latent-distribution-test-statistic">
<h2>For Unmatched Networks - Latent Distribution Test Statistic<a class="headerlink" href="#for-unmatched-networks-latent-distribution-test-statistic" title="Permalink to this headline">¶</a></h2>
<p>The <a class="reference external" href="https://microsoft.github.io/graspologic/latest/reference/reference/inference.html#graspologic.inference.latent_distribution_test">latent distribution test</a> is a two-sample hypothesis test for whether or not two graphs have the same distributions of latent positions. This test does not assume any known matching, as it can use alignment techniques such as <a class="reference external" href="https://microsoft.github.io/graspologic/latest/reference/reference/align.html#graspologic.align.SeedlessProcrustes">seedless-procrustes</a> to find the best orthogonal alignment between two embeddings.</p>
<p>First we generate <a class="reference external" href="https://microsoft.github.io/graspologic/latest/reference/reference/embed.html#graspologic.embed.AdjacencySpectralEmbed">adjacency spectral embeddings</a> of the largest connected component of the unmatched graphs, then calculate the latent distribution test statistic for every pair to construct the dissimilarity matrix using parameters shown in the code below.</p>
<div class="cell tag_hide-input docutils container">
<div class="cell_input docutils container">
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># HIDE CODE</span>
<span class="kn">from</span> <span class="nn">graspologic.inference</span> <span class="kn">import</span> <span class="n">latent_distribution_test</span>
<span class="kn">from</span> <span class="nn">graspologic.embed</span> <span class="kn">import</span> <span class="n">AdjacencySpectralEmbed</span>
<span class="kn">from</span> <span class="nn">graspologic.utils</span> <span class="kn">import</span> <span class="n">largest_connected_component</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>

<span class="n">ase_graphs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">graph</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">graphs</span><span class="p">):</span>
    <span class="n">lcc_graph</span> <span class="o">=</span> <span class="n">largest_connected_component</span><span class="p">(</span><span class="n">graph</span><span class="p">)</span>
    <span class="n">ase_graph</span> <span class="o">=</span> <span class="n">AdjacencySpectralEmbed</span><span class="p">(</span><span class="n">n_components</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">lcc_graph</span><span class="p">)</span>
    <span class="n">ase_graphs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ase_graph</span><span class="p">)</span>

<span class="c1"># calculate alignments</span>
<span class="n">latent_dissim</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="nb">len</span><span class="p">(</span><span class="n">ase_graphs</span><span class="p">),</span> <span class="nb">len</span><span class="p">(</span><span class="n">ase_graphs</span><span class="p">)))</span>
<span class="n">Qs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ase_graphs</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)):</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">j</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">ase_graphs</span><span class="p">)):</span>
        <span class="n">statistic</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">misc_dict</span> <span class="o">=</span> <span class="n">latent_distribution_test</span><span class="p">(</span><span class="n">ase_graphs</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">ase_graphs</span><span class="p">[</span><span class="n">j</span><span class="p">],</span> <span class="n">test</span><span class="o">=</span><span class="s1">&#39;mgc&#39;</span><span class="p">,</span> <span class="n">metric</span><span class="o">=</span><span class="s1">&#39;euclidean&#39;</span><span class="p">,</span> \
            <span class="n">n_bootstraps</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">align_type</span><span class="o">=</span><span class="s1">&#39;seedless_procrustes&#39;</span><span class="p">,</span> <span class="n">input_graph</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">latent_dissim</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">statistic</span>
        <span class="n">Qs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">misc_dict</span><span class="p">[</span><span class="s1">&#39;Q&#39;</span><span class="p">])</span>

<span class="c1"># scale values from 0 to 1 and make matrix symmetric</span>
<span class="n">scaled_latent_dissim</span> <span class="o">=</span> <span class="n">latent_dissim</span> <span class="o">/</span> <span class="n">np</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">latent_dissim</span><span class="p">)</span>
<span class="n">scaled_latent_dissim</span> <span class="o">=</span> <span class="n">scaled_latent_dissim</span> <span class="o">+</span> <span class="n">scaled_latent_dissim</span><span class="o">.</span><span class="n">T</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="references">
<h1>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h1>
<div class="docutils container" id="id8">
<dl class="citation">
<dt class="label" id="id9"><span class="brackets"><a class="fn-backref" href="#id6">1</a></span></dt>
<dd><p>Laura E Suárez, Yossi Yovel, Martijin P Van den Heuvel, Olaf Sporns, Yaniv Assaf, Guillaume Lajoie, and Bratislav Misic. A connectomics-based taxonomy of mammals. <em>bioRxiv</em>, 03 2022. <a class="reference external" href="https://doi.org/10.1101/2022.03.11.483995">doi:10.1101/2022.03.11.483995</a>.</p>
</dd>
<dt class="label" id="id10"><span class="brackets"><a class="fn-backref" href="#id1">2</a></span></dt>
<dd><p>Eric W. Weisstein. L^1-norm. Accessed May 16, 2022 [Online]. URL: <a class="reference external" href="https://mathworld.wolfram.com/L1-Norm.html">https://mathworld.wolfram.com/L1-Norm.html</a>.</p>
</dd>
<dt class="label" id="id11"><span class="brackets"><a class="fn-backref" href="#id2">3</a></span></dt>
<dd><p>Eric W. Weisstein. L^2-norm. Accessed May 16, 2022 [Online]. URL: <a class="reference external" href="https://mathworld.wolfram.com/L2-Norm.html">https://mathworld.wolfram.com/L2-Norm.html</a>.</p>
</dd>
<dt class="label" id="id12"><span class="brackets">4</span><span class="fn-backref">(<a href="#id3">1</a>,<a href="#id4">2</a>)</span></dt>
<dd><p>Eric W. Weisstein. Matrix norm. Accessed May 16, 2022 [Online]. URL: <a class="reference external" href="https://mathworld.wolfram.com/MatrixNorm.html">https://mathworld.wolfram.com/MatrixNorm.html</a>.</p>
</dd>
<dt class="label" id="id14"><span class="brackets"><a class="fn-backref" href="#id5">5</a></span></dt>
<dd><p>K. Ch. Das. The laplacian spectrum of a graph. <em>Computers &amp; Mathematics with Applications</em>, 48(5-6):715–724, 09 2004. <a class="reference external" href="https://doi.org/10.1016/j.camwa.2004.05.005">doi:10.1016/j.camwa.2004.05.005</a>.</p>
</dd>
<dt class="label" id="id13"><span class="brackets"><a class="fn-backref" href="#id7">6</a></span></dt>
<dd><p>Gene H. Golub and Charles F. Van Loan. <em>Matrix Computations</em>. The Johns Hopkins University Press, Baltimore, MD, 1985.</p>
</dd>
</dl>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="data_description.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Datasets</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="mice_discriminability.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Matched Networks</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Da In (Diane) Lee<br/>
    
        &copy; Copyright 2021.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>